6 every deployment replaces the bridge certificate

pudkrong opened the pull request:
In development environment, the generated certificates are saved into `certificates` folder and will be reused if they are existing. Therefore, there are no change in CDK. 

However, in deployment workflow, the `certificates` folder never exists. So, the certificates are created every time and cause changes in CDK.

In this PR, I have backup the certificates in SSM and restore them if exists.

pudkrong committed:
feat: use lambda source

pudkrong committed:
feat: backup MQTT certificates into SSM

pudkrong committed:
test: improve mock function

pudkrong committed:
feat: add clean backup certificates cli

pudkrong committed:
feat: add debug log

pudkrong committed:
feat: do not back restored certificates

coderbyheart commented on bridge/certificatesSSM.ts:
```diff
@@ -0,0 +1,83 @@
+	const parameters = await getSettingsOptional<Record<string, string>, null>({
+		ssm,
+		stackName: STACK_NAME,
+		scope: 'codebuild',
```
We don't use [codebuild](https://docs.aws.amazon.com/codebuild/latest/userguide/welcome.html).

coderbyheart commented on bridge/certificatesSSM.ts:
```diff
@@ -0,0 +1,83 @@
+				hasRestored = true
+			}
+		}
+	}
```
Instead of inventing another way of nesting parameters, use SSM Parameter nesting.

coderbyheart commented on bridge/certificatesSSM.ts:
```diff
@@ -0,0 +1,83 @@
+}): Promise<boolean> => {
+	let hasRestored = false
+
+	const parameters = await getSettingsOptional<Record<string, string>, null>({
+		ssm,
+		stackName: STACK_NAME,
+		scope: 'codebuild',
+		system: 'stack',
+	})(null)
```
```suggestion
	const parameters = await getSettingsOptional<Record<string, string>, null>({
		ssm,
		stackName: STACK_NAME,
		scope: 'config',
		system: 'mqttBridge',
	})(null)
```

coderbyheart commented:
I've changed the Settings API in https://github.com/hello-nrfcloud/backend/commit/34e04b92ac9d98554678082b0c6066e6d114ab1d ... so that only used scope/system combinations are documented which also makes it easier to introduce new scopes.

coderbyheart commented:
New scope:

```typescript
export enum Scope {
...
	NRFCLOUD_BRIDGE_CONFIG = 'nrfcloud/bridge'
...
}
```

pudkrong committed:
feat: convert SSM to nested object

coderbyheart commented on bridge/certificatesSSM.ts:
**suggestion:** If you separate the concern of writing and reading files from the writing and reading from SSM in this file, testing becomes easier.

coderbyheart commented on util/settings.spec.ts:
```diff
@@ -29,3 +34,73 @@ describe('settingsPath()', () => {
+	it('should return the object with same scope', async () => {
+		const returnedValues = [
+			{
+				Name: `/${'STACK_NAME'}/${Scope.STACK_CONFIG}/key1`,
```
```suggestion
				Name: `/hello-nrfcloud/stack/context/key1`,
```

coderbyheart replied:
Avoid using placeholders in mock data, so it's more clear what the real format is supposed to be.

pudkrong committed:
test: refactor

coderbyheart committed:
refactor: remove backupCertificatesToSSM

Use the existing settings API instead
and extract the reading of a map of files
to a separate function

coderbyheart commented on bridge/certificatesSSM.ts:
```diff
@@ -0,0 +1,46 @@
+	if (parameterNamePrefix in parameters) {
+		const certificatesInSSM = parameters[parameterNamePrefix]
+		for (const key in certificates) {
+			if (typeof certificatesInSSM === 'object' && key in certificatesInSSM) {
```
Note: this is a hack, it does not match the type of parameters.

coderbyheart committed:
refactor: remove restoreCertificatesFromSSM

Use the existing settings API instead
and extract the reading of a map of files
to a separate function

This also fixes the issue with the proposed implementation
that goes around the settings API types

coderbyheart committed:
refactor: remove hidden nesting support

coderbyheart committed:
fix: map from keys to path names and validate restored object

Previously the certificate map was not validated
whether it is complete
